### DISCUSSION TITLE
Using pyboard.py and ESP32 WIFI

### DISCUSSION THREAD

**Discussion by A622266** [micropython/micropython#9488] [Hardware & Peripherals] [answered]
I have an ESP32 connected to my home network at 192.168.1.3
I'm wondering if I can use the script pyboard.py to interact with the board over WIFI, instead of USB.  I think it may be possible, since the documentation says the device argument can be  *the serial device or the IP address of the pyboard*.

I tried changing the environment variable:
```py
export PYBOARD_DEVICE=ip:192.168.1.3
```
When I run the script:
```py
python pyboard.py main.py
```
to try to run local file main.py on the ESP32, I get an error:  *failed to access ip:192.168.1.3*
I don't see any examples of using an ip address as the device argument in the documentation.  Can anyone give advice? 

Regards
Andy

**Comment by jimmo** [comment, marked as answer]
This is not currently supported on the ESP32. I think just the CC3200 port supports it (via its built-in telnet server).

In the meantime webrepl is the best tool to use to access the esp32 over thr network, and we plan to add webrepl support to mpremote/pyboard.

**Reply by A622266** [reply]
thanks.  Is mpremote intended as a replacement for pyboard.py?  It seems to have similar features.

**Reply by cwalther** [reply]
> we plan to add webrepl support to mpremote/pyboard

Interesting. When I tried that a long time ago, I tripped over the design flaw that the WebREPL protocol is not binary-safe (it puts arbitrary input unencoded into WebSocket text messages, which are only allowed to contain UTF-8). Do you have any plans yet on how to fix the WebREPL protocol? I gave up at some point because I found it hard to extend backward-compatibly.

**Reply by dpgeorge** [reply]
WebSocket supports binary frames, see eg https://www.rfc-editor.org/rfc/rfc6455#page-29

**Reply by jimmo** [reply]
> thanks. Is mpremote intended as a replacement for pyboard.py? It seems to have similar features.

@A622266 no, `mpremote` uses `pyboard.py` internally. pyboard.py is the low-level tool (and can be used programatically from Python too), whereas `mpremote` is the user-facing tool (and it's published on PyPI too).

> Interesting. When I tried that a long time ago, I tripped over the design flaw that the WebREPL protocol is not binary-safe (it puts arbitrary input unencoded into WebSocket text messages, which are only allowed to contain UTF-8). Do you have any plans yet on how to fix the WebREPL protocol? I gave up at some point because I found it hard to extend backward-compatibly.

@cwalther Yes, I'm not sure how we're going to solve this. But like you say, it's broken as it is at the moment. I'm not sure how many people were using webrepl until recently anyway because the http/https thing (which is now fixed).
I think the fix will likely be to switch to "new" protocol for all firmware built past some date, make micropython.org/webrepl support both, and make mpremote only use the new one.

**Reply by cwalther** [reply]
Sounds good. By the way, I haven’t checked yet what CircuitPython is doing for their new web workflow, but maybe there’s some inspiration there (or even opportunity for compatibility).

**Comment by robert-hh** [comment]
For fuile access using wifi, you could as well use the ftp server. https://github.com/robert-hh/FTP-Server-for-ESP8266-ESP32-and-PYBD
